ONSdigital/eq-survey-runner
# Conditionally render and set cookies/cookiesbanner

### What is the context of this PR?

Adds the option to accept all cookies and select individual cookies of the users choice. 

A cookie banner has been added to the top *any* page and is removed once a user accepts all cookies or modifies their preferences.

To modify their preferences a `/cookies-settings` page has been added which only gives the option to remove google analytics at this point. From what I can tell, the other options provided by the ONS cookie JS are relevant to `eq-survey-runner`. However I could be wrong.

At this point 99% of the functionality is complete with the supporting tests, the remaining work to be done is content and quality of life checks. Most of the templates have been taken from elsewhere in the ONS and some of the content doesn't necessarily reflect that.

One content issue I have noticed so far is that once the user has selected the cookies they want to allow. There is no way for them to move off of the settings page. At the moment I've left a placeholder for a link but not quite sure where to direct the link to.

The only major concern I have with this work is I have removed the google analytics scripts being added to the templates when the user has chosen to do so. However, there is an `analytics.js` script being brought in by the ONS cdn and I don't think I can modify it? There is a local version of the same file within runner but it doesn't seem to be used.

One minor concern and I'm not actually sure how to fix yet is cookies only being enabled on moving to a new page or after a refresh. This means after accepting all cookies, the analytics scripts will only be pulled in on moving to the next page.

### How to review 
This might take a while
#### Test one
1. Getting runner running locally (using docker provides the GA variables)
2. Launch a questionnaire and check that the cookie banner is present (check several different types of survey)
3. Open the chrome dev tools > application > cookies
4. Click the accept all cookies button and notice the `ons_cookies_message_displayed` cookie get added
5. Take a note of the value of the cookie policy
6. The banner should be replaced with another message with a hide able button
7. Hit refresh
8. The banner should now be gone
9. Clear the cookies

#### Test two
1. Launch a survey
2. Click the Set cookie preferences button
3. Read through the content to make sure if makes sense
4. The page is different from the ONS design system one because as far as I can tell we don't have that many cookies
5. Check the cookies
6. Press yes to the GA cookie and submit
7. Cookie confirmation banner should appear (need to decide where to send user after this)
8. Refresh the page and the options should represent the selected values

Also need to run the `js` integration tests to make sure they pass. They don't come up in the PR.

To run them use `./scripts/run_tests_functional.sh`

I won't be in the rest of the week but if there are any questions during reviews on this, message me on slack or leave a comment in the PR and I'll respond whilst I can.
